src/operator/tensor/elemwise_unary_op_basic.cc

```diff
@@ -121,7 +121,20 @@ The storage type of ``sigmoid`` output is always dense
 .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseOut{"_backward_sigmoid"});
 
 MXNET_OPERATOR_REGISTER_BINARY_WITH_SPARSE_CPU(_backward_sigmoid,
-                                               unary_bwd<mshadow_op::sigmoid_grad>);
+                                               unary_bwd<mshadow_op::sigmoid_grad>)
+.set_attr<nnvm::FGradient>("FGradient",
+  [](const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ograds) {
+    auto fx = nnvm::NodeEntry{n->inputs[1]};
+    auto gx_ograd = nnvm::NodeEntry{n};
+
+    std::vector<nnvm::NodeEntry> ret;
+       
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad_inp",
+                             {ograds[0], gx_ograd}, nullptr, &n));
+    ret.emplace_back(MakeNode("elemwise_mul", n->attrs.name + "_backward_grad_grad",
```

Comment:
Shouldn't this be ograds[0] * f'(x) ?

Let f(x) = sigmoid(x)
xgrad = dL/dx
ygrad = dL/dy
We know dydx = y*(1-y) 
In the first backward pass:
```
xgrad = dydx * ygrad

n->inputs[0] ---> ygrad
n->inputs[1] ---> y = f(x)
nnvm::NodeEntry{n} ---> xgrad
```

According to the multiplier rule, in the second backward pass:
```
_backward_grad_grad = d2L/dx2 * f'(x) = ograds[0] * dydx
_backward_grad_grad_inp = d2L/dx2 * ygrad * f''(x) = ograds[0] * n->input[0] * f''(x)
```

Reply:
That is indeed the surprising part, I have asserted from Python that `n->inputs[1]` is `f(x) * f'(x)` surprisingly and if I remember correctly `n->inputs[0]` as well as `nnvm::NodeEntry{n}` are different from what we expect them to be.

Reply:
I got your question now. I got the same result `f(x) * f'(x) * ograds` locally. Need to dig in more...

Reply:
I dumped out the computation graph. If you want to try, refer to my local branch: https://github.com/apeforest/incubator-mxnet/blob/develop/higher_order_grad/src/imperative/imperative.cc#L509
```
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:300: node 0 var
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:300: node 1 var
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:302: node 2 sigmoid
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:306: 		input 1: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:312: 		output 2: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:302: node 3 _backward_sigmoid
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:306: 		input 0: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:306: 		input 2: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:312: 		output 3: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:300: node 4 var
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:302: node 5 elemwise_mul
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:306: 		input 4: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:306: 		input 2: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:312: 		output 5: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:302: node 6 _backward_sigmoid
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:306: 		input 5: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:306: 		input 2: [3] (0 KB) ->
[16:37:15] /Users/lnyuan/work/mxnet/src/executor/../common/exec_utils.h:312: 		output 6: [3] (0 KB) ->
```

So it seems if the backward operator takes output of the forward node instead of input, the backward graph may be created differently with some chain rule applied already.

Reply:
Oh Thanks. Will try to look into it. Thanks for the pointer.